File: java/SmallCube.java
package rubikscube;

import javafx.scene.Group;
import javafx.scene.paint.Color;
import javafx.scene.shape.Line;
import javafx.scene.shape.Rectangle;
import javafx.scene.transform.Rotate;
import javafx.scene.transform.Translate;

public class SmallCube extends Group{
    
    public static final int FRONT = 0;
    public static final int TOP = 1;
    public static final int BOTTOM = 2;
    public static final int LEFT = 3;
    public static final int RIGHT = 4;
    public static final int BACK = 5;
    
    private int[] positionIndices;
    public boolean isDraggable;
    
    public Color[] colors;
    private Rectangle[] surfaces;

    public SmallCube(int s){
        positionIndices = new int[3];
        isDraggable = false;
        colors = new Color[6];
        
        surfaces = new Rectangle[6];
        for(int i=0; i<6; i++){
            Rectangle r = new Rectangle(0, 0, s, s);
            surfaces[i] = r;
        }
        
        //0: front
        
        
        //1: top
        surfaces[1].getTransforms().addAll(
            new Rotate(90, Rotate.X_AXIS));
        
        //2: bottom
        surfaces[2].getTransforms().addAll(
            new Rotate(90, Rotate.X_AXIS),
            new Translate(0, 0, -s));
        
        //3: left
        surfaces[3].getTransforms().addAll(
            new Rotate(270, Rotate.Y_AXIS));
        
        //4: right
        surfaces[4].getTransforms().addAll(
            new Rotate(270, Rotate.Y_AXIS),
            new Translate(0, 0, -s));
        
        //5: back
        surfaces[5].getTransforms().addAll(
            new Translate(0, 0, s));
        
        
        Line[] lines = new Line[12];
        
        lines[0] = new Line(0, 0, s, 0);
        lines[1] = new Line(0, 0, 0, s);
        lines[2] = new Line(0, s, s, s);

        lines[3] = new Line(0, 0, s, 0);
        lines[4] = new Line(0, 0, 0, s);
        lines[5] = new Line(0, s, s, s);
        for(int i=3; i<6; i++)
            lines[i].getTransforms().addAll(
                new Rotate(90, Rotate.Y_AXIS),
                new Translate(-s, 0, 0));
        
        lines[6] = new Line(0, 0, s, 0);
        lines[7] = new Line(0, 0, 0, s);
        lines[8] = new Line(0, s, s, s);
        for(int i=6; i<9; i++)
            lines[i].getTransforms().addAll(
                new Rotate(180, Rotate.Y_AXIS),
                new Translate(-s, 0, -s));
        
        lines[9] = new Line(0, 0, s, 0);
        lines[10] = new Line(0, 0, 0, s);
        lines[11] = new Line(0, s, s, s);
        for(int i=9; i<12; i++)
            lines[i].getTransforms().addAll(
                new Rotate(270, Rotate.Y_AXIS),
                new Translate(0, 0, -s));
        
        for(int i=0; i<12; i++) lines[i].setFill(Color.BLACK);
        
        getChildren().addAll(surfaces);
        getChildren().addAll(lines);
    }
    
    public int[] getPositionIndices(){
        return positionIndices;
    }
    
    public void setPositionIndices(int[] indices){
        positionIndices = indices;
    }
    
    public Color getColor(int surface){
        return (Color) surfaces[surface].getFill();
    }
    
    public void setColor(int surface, Color c, boolean saveColor){
        surfaces[surface].setFill(c);
        if(saveColor) colors[surface] = c;
    }
    
    public void setColor(Color c, boolean saveColor){
        for(int i=0; i<6; i++) setColor(i, c, saveColor);
    }
    
    public void refreshColors(){
        for(int i=0; i<6; i++){
            if(colors[i] != null) surfaces[i].setFill(colors[i]);
        }
    }
    
    @Override
    public String toString(){
        return "(" + positionIndices[0] + ", " + positionIndices[1] + ", " + positionIndices[2] + ")";
    }
    
}

File: javascript/smallCube.js
class SmallCube{

	//ranges from (0,0,0) to (width, width, -width)
	constructor(width, scene, transforms){
		this.positionIndices = [];
        this.isDraggable = false;
        this.colors = [];
        
        this.surfaces = multiArray([6]);
		
		//front: from (0,0,0) to (width, width, 0)
		/*this.surfaces[0] = new Surface(width, width, this);
		this.surfaces[0].translate(-width, -width, -3*width/2);
		this.surfaces[0].rotate(0, Math.PI, 0);
		
		//back: from (0,0,-width) to (width, width, -width)
		this.surfaces[1] = new Surface(width, width, this);
		this.surfaces[1].translate(-width, -width, -width/2);
		
		//left: from (0,0,0) to (0, width, -width)
		this.surfaces[2] = new Surface(width, width, this);
		this.surfaces[2].translate(-3*width/2, -width, -width);
		this.surfaces[2].rotate(0, -Math.PI/2, 0);
		
		//right: from (1,0,0) to (1, width, -width)
		this.surfaces[3] = new Surface(width, width, this);
		this.surfaces[3].translate(-width/2, -width, -width);
		this.surfaces[3].rotate(0, -Math.PI/2, 0);
		
		//top: from (0,width,0) to (width, width, -width)
		this.surfaces[4] = new Surface(width, width, this);
		this.surfaces[4].translate(-width, -3*width/2, -width);
		this.surfaces[4].rotate(-Math.PI/2, 0, 0);
		
		//bottom: from (0,0,0) to (width, 0, -width)
		this.surfaces[5] = new Surface(width, width, this);
		this.surfaces[5].translate(-width, -width/2, -width);
		this.surfaces[5].rotate(Math.PI/2, 0, 0);*/
		
		//front: from (0,0,0) to (width, width, 0)
		var t = copyArray(transforms);
		t.push(new Translate(-width/2, -3*width/2, -3*width/2));
		t.push(new Rotate(Math.PI, axis.y));
		this.surfaces[0] = new Surface(width, width, this, t);
		
		//back: from (0,0,-width) to (width, width, -width)
		t = copyArray(transforms);
		t.push(new Translate(-width/2, -3*width/2, -width/2));
		t.push(new Rotate(Math.PI, axis.y));
		this.surfaces[1] = new Surface(width, width, this, t);
		
		//left: from (0,0,0) to (0, width, -width)
		t = copyArray(transforms);
		t.push(new Translate(-3*width/2, -3*width/2, -3*width/2));
		t.push(new Rotate(-Math.PI/2, axis.y));
		this.surfaces[2] = new Surface(width, width, this, t);
		
		//right: from (1,0,0) to (1, width, -width)
		t = copyArray(transforms);
		t.push(new Translate(-width/2, -3*width/2, -3*width/2));
		t.push(new Rotate(-Math.PI/2, axis.y));
		this.surfaces[3] = new Surface(width, width, this, t);
		
		//top: from (0,width,0) to (width, width, -width)
		t = copyArray(transforms);
		t.push(new Translate(-3*width/2, -3*width/2, -3*width/2));
		t.push(new Rotate(Math.PI/2, axis.x));
		this.surfaces[4] = new Surface(width, width, this, t);
		
		//bottom: from (0,0,0) to (width, 0, -width)
		t = copyArray(transforms);
		t.push(new Translate(-3*width/2, -width/2, -3*width/2));
		t.push(new Rotate(Math.PI/2, axis.x));
		this.surfaces[5] = new Surface(width, width, this, t);
		
		
		this.group = new THREE.Group();
		for(var i=0; i<this.surfaces.length; i++){
			this.surfaces[i].addToGroup(this.group);
		}
	}
	
	update(){
		for(var i=0; i<this.surfaces.length; i++){
			this.surfaces[i].update();
		}
	}
	
	addToGroup(group){
		group.add(this.group);
	}
	
	
	/*rotateX(angle){
		this.group.rotation.x = angle;
	}
	
	rotateY(angle){
		this.group.rotation.y = angle;
	}
	
	rotateZ(angle){
		this.group.rotation.z = angle;
	}*/
	
	/*rotate(rotateX, rotateY, rotateZ){
		this.group.rotateX(rotateX);
		this.group.rotateY(rotateY);
		this.group.rotateZ(rotateZ);
	}
	
	translate(translateX, translateY, translateZ){
		this.group.translateX(translateX);
		this.group.translateY(translateY);
		this.group.translateZ(translateZ);
	}*/
	
	getPositionIndices(){
        return this.positionIndices;
    }
    
    setPositionIndices(indices){
        this.positionIndices = indices;
    }
    
    getColor(surfaceIndex){
        return this.colors[surfaceIndex];
    }

	setColor(surfaceIndex, surfaceColor, saveColor){
		this.surfaces[surfaceIndex].setColor(surfaceColor);
		if(saveColor) this.colors[surfaceIndex] = surfaceColor;
	}

    /*setColor(surfaceColor, saveColor){
        for(var i=0; i<6; i++) this.setColorAt(i, surfaceColor, saveColor);
    }*/
	
    refreshColors(){
        for(var i=0; i<6; i++){
            if(colors[i]) this.surfaces[i].setColor(this.colors[i]);
        }
    }
    
    toString(){
        return "(" + this.positionIndices[0] + ", " + this.positionIndices[1] + ", " + this.positionIndices[2] + ")";
    }
	
	on(eventName, eventFunction){
		this.group.on(eventName, eventFunction);
	}

}